### ISSUE TITLE
VideoPlayer throws error when it loads video without sound

### ISSUE THREAD

**Issue by eaglecat** [godotengine/godot#5133]
here's example godot code
```
var re = ResourceLoader.load("res://video.ogv")
get_node("VideoPlayer").set_stream(re)
get_node("VideoPlayer").play()
```

if video.ogv do not contains 'sound', VideoPlayer throws error.
else if it contains any little sound, works well.

![image](http://i.imgur.com/wvxTDwT.png)

**Comment by vnen** [comment]
Confirmed by @Dillybob92 in #4878.

**Comment by reduz** [comment]
uh it's super difficult to get this to work, Theora demuxer code is hell.  Leave it open but don't count on it being fixed anytime soon. Consider using a video with silent sound.

**Comment by zaps166** [comment]
Btw. isn't FFmpeg integration a good idea? It can play any container and any codec (audio and video and also a still image). It is well optimized, multithreaded, stable and its API is simple.
You can configure FFmpeg and choose only necessary protocols, formats, decoders to reduce library size.
What do you think (as a future feature)?

**Comment by zaps166** [comment]
@reduz I don't know side effects of this patch, but it seems to be working properly on video with and without sound :)

Changes: Don't wait for audio data when vorbis data doesn't exists. Don't init the resampler with 0 audio channels.

```diff
diff --git a/drivers/theora/video_stream_theora.cpp b/drivers/theora/video_stream_theora.cpp
index e577c3f..fc296e2 100644
--- a/drivers/theora/video_stream_theora.cpp
+++ b/drivers/theora/video_stream_theora.cpp
@@ -513,7 +513,7 @@ void VideoStreamPlaybackTheora::update(float p_delta) {
        }
 
        bool frame_done=false;
-       bool audio_done=false;
+       bool audio_done=!vorbis_p;
 
        bool theora_done=false;
 
diff --git a/scene/gui/video_player.cpp b/scene/gui/video_player.cpp
index 26227d6..bb87b83 100644
--- a/scene/gui/video_player.cpp
+++ b/scene/gui/video_player.cpp
@@ -209,7 +209,10 @@ void VideoPlayer::set_stream(const Ref<VideoStream> &p_stream) {
                texture=playback->get_texture();
 
                AudioServer::get_singleton()->lock();
-               resampler.setup(playback->get_channels(),playback->get_mix_rate(),server_mix_rate,buffering_ms,0);
+               if (playback->get_channels() > 0)
+                       resampler.setup(playback->get_channels(),playback->get_mix_rate(),server_mix_rate,buffering_ms,0);
+               else
+                       resampler.clear();
                AudioServer::get_singleton()->unlock();
                playback->set_mix_callback(_audio_mix_callback,this);
 
```

**Comment by eaglecat** [comment]
> Btw. isn't FFmpeg integration a good idea?
@zaps166  oh wow. exactly what I always wanted !!

**Comment by SuperUserNameMan** [comment]
@zaps166 : if integrating ffmpeg, how large would it make godot.exe ?

**Comment by zaps166** [comment]
It depends on FFmpeg compilation. You can enable only necessary codecs and formats to decrease the size. It can be linked statically (less size, one file). FFmpeg with all formats, all codecs and all protocols (DLLs) has **~12MiB** (3.0.2 version) on Windows 32bit.
If someone will decide to integrate it then all existing audio and video codecs in Godot should be removed to decrease size and increase performance, because FFmpeg can handle all existing codecs in Godot better. E.g. libvorbis is much slower than FFmpeg built-in Vorbis decoder :)
FFmpeg has also quite good and fast audio resampler and video scaler.
The minimal build which I've made now with only a few codecs and formats can take **~4MiB**. Without theora, libvorbis, opus, and other existing codecs (and mybe resampler, etc) in Godot it won't increase the size much.
It also can decode still images like JPEG/PNG. but I've never used FFmpeg for still images :)
FFmpeg has its own protocols support (http, https, file, rtsp, rtmp, ftp, etc...) - they can be compiled-in and used (also in other Godot parts), or they might be disabled during compilation. Then custom IO reader can be created which will use Gotots' file and network functions.

See also: https://github.com/zaps166/QMPlay2/tree/master/src/modules/FFmpeg

**Comment by SuperUserNameMan** [comment]
hmm, interesting.

if i'm not wrong, ffmepg license is GPL or LGPL. 
Would it be compatible with godo's license ?

If not, instead of integrating into godot's core, could it be an external module ? and would it be compatible with GPL or LGPL ?

**Comment by zaps166** [comment]
The FFmpeg license is GPL and/or LGPL depends on the `configure` command line. Changing the license to GPL can disable some encoders (encoders are not needed for Godot), maybe filters or maybe some decoders.

About a license please ask developers - I don't know :)

**Comment by zaps166** [comment]
If other people are interested in FFmpeg integration maybe it is better to create a new discussion issue?

This issue can be fixed by https://github.com/godotengine/godot/issues/5133#issuecomment-225399974, but I don't know if it is 100% correct (I tested it and it worked on OGG and Theora video with and without sound). This just prevents infinity loop which waits for audio data which doesn't exists and silences the error due to 0 channels. I'll look into it again and I'll do a PR later.

**Comment by akien-mga** [comment]
Minimal reproducer from @Dillybob92 in #4878: https://github.com/godotengine/godot/files/288424/ogv.bug.zip

**Comment by zaps166** [comment]
I've made a PR (2.0 branch, because it is only a bugfix). Please test (for me works properly with attached video).

**Comment by akien-mga** [comment]
I prefer bugfix PRs made against master, as the bug affects both master and 2.0. It's better for users if I forget to cherry-pick the master commit for 2.0 than if I forget to cherry-pick the 2.0 commit for master ;)

**Comment by zaps166** [comment]
Ok, so my next bugfix PR will be in `master` :)

**Comment by vnen** [comment]
This was fixed by #5224.
